cli/azd/pkg/ext/hooks.go

```diff
@@ -0,0 +1,379 @@
+	// Delete any temporary inline scripts after execution
+	defer func() {
+		if scriptConfig.Location == ScriptLocationInline {
+			os.Remove(scriptConfig.Path)
+		}
+	}()
+
+	script, err := h.GetScript(scriptConfig)
+	if err != nil {
+		return err
+	}
+
+	formatter := h.console.GetFormatter()
+	consoleInteractive := formatter == nil || formatter.Kind() == output.NoneFormat
+	scriptInteractive := consoleInteractive && scriptConfig.Interactive
+
+	// When running in an interactive terminal broadcast a message to the dev to remind them that custom hooks are running.
+	if consoleInteractive {
+		h.console.Message(
+			ctx,
+			output.WithBold(
+				fmt.Sprintf(
+					"Executing %s command hook => %s",
+					output.WithHighLightFormat(scriptConfig.Name),
+					output.WithHighLightFormat(scriptConfig.Path),
+				),
+			),
+		)
+	}
+
+	log.Printf("Executing script '%s'\n", scriptConfig.Path)
+	_, err = script.Execute(ctx, scriptConfig.Path, scriptInteractive)
+	if err != nil {
+		execErr := fmt.Errorf("failed executing script '%s' : %w", scriptConfig.Path, err)
+
+		// If an error occurred log the failure but continue
+		if scriptConfig.ContinueOnError {
+			h.console.Message(ctx, output.WithBold(output.WithWarningFormat("WARNING: %s", execErr.Error())))
+			h.console.Message(
+				ctx,
+				output.WithWarningFormat("'%s' script has been configured to continue on error.", scriptConfig.Name),
+			)
+			log.Println(execErr.Error())
+		} else {
+			return execErr
+		}
+	}
+
+	return nil
+}
+
+func inferScriptTypeFromFilePath(path string) (ScriptType, error) {
+	fileExtension := filepath.Ext(path)
+	switch fileExtension {
+	case ".sh":
+		return ScriptTypeBash, nil
+	case ".ps1":
+		return ScriptTypePowershell, nil
+	default:
+		return "", fmt.Errorf(
+			"script with file extension '%s' is not valid. Only '.sh' and '.ps1' are supported",
+			fileExtension,
+		)
+	}
+}
+
+func createTempScript(scriptConfig *ScriptConfig) (string, error) {
+	var ext string
+	scriptHeader := []string{}
+
+	switch scriptConfig.Type {
+	case ScriptTypeBash:
+		ext = "sh"
+		scriptHeader = []string{
+			"#!/bin/sh",
+		}
+	case ScriptTypePowershell:
+		ext = "ps1"
+	}
+
+	// Creates .azure/hooks directory if it doesn't already exist
+	// In the future any scripts with names like "predeploy.sh" or similar would
+	// automatically be invoked base on our command hook naming convention
+	directory := filepath.Join(".azure", "hooks")
+	_, err := os.Stat(directory)
+	if err != nil {
+		err := os.MkdirAll(directory, osutil.PermissionDirectory)
+		if err != nil {
+			return "", fmt.Errorf("failed creating command hooks directory, %w", err)
+		}
+	}
+
+	// Write the temporary script file to .azure/hooks folder
+	file, err := os.CreateTemp(directory, fmt.Sprintf("%s-*.%s", scriptConfig.Name, ext))
+	if err != nil {
+		return "", fmt.Errorf("failed creating command hook file: %w", err)
+	}
+
+	defer file.Close()
+
+	scriptBuilder := strings.Builder{}
```

Comment:
Seems like we can use `bufio.NewWriter` instead for max efficiency

Reply:
Will look into this.

Reply:
Going to hold off on this and not looking to overly optimize at this point.